File: packages/mirascript/src/vm/lib/global/sequence/sort.ts
import { $Call, $ToNumber } from '../../../operations.js';
import type { VmAny, VmConst, VmValue } from '../../../types/index.js';
import { VmLib, expectArray, expectCallable } from '../../_helpers.js';

/** 默认比较 */
function defaultCompare(a: VmValue = null, b: VmValue = null): number {
    if (Object.is(a, b)) return 0;
    if ((typeof a == 'string' || a == null) && (typeof b == 'string' || b == null)) {
        a ??= '';
        b ??= '';
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
    }
    // nan is treated as 0
    const an = $ToNumber(a) || 0;
    const bn = $ToNumber(b) || 0;
    if (an < bn) return -1;
    if (an > bn) return 1;
    return 0;
}

/** 获取比较函数 */
function cmp(comparator: VmAny, recovered: VmValue): typeof defaultCompare {
    if (comparator == null) return defaultCompare;
    expectCallable('comparator', comparator, recovered);
    return (a: VmValue = null, b: VmValue = null) => {
        const ret = $Call(comparator, [a, b]);
        return $ToNumber(ret);
    };
}

export const sort = VmLib(
    (data, comparator) => {
        expectArray('data', data, null);
        const compare = cmp(comparator, data);
        const arr: VmConst[] = [];
        for (const v of data) {
            arr.push(v ?? null);
        }
        arr.sort(compare);
        return arr;
    },
    {
        summary: '对数组中的元素进行排序，并返回排序后的结果',
        params: {
            data: '要排序的数组',
            comparator: '用于比较两个元素的函数，返回一个数字，表示它们的相对顺序，默认按升序排列',
        },
        paramsType: {
            data: 'array',
            comparator: 'fn(a: any, b: any) -> number',
        },
        returnsType: 'array',
        examples: ['sort(["c", "a", "b"]) // ["a", "b", "c"]'],
    },
);

export const sort_by = VmLib(
    (data, key_fn, comparator) => {
        expectArray('data', data, null);
        expectCallable('key_fn', key_fn, data);
        const compare = cmp(comparator, data);
        const arr: Array<{ o: VmConst; k: VmValue }> = [];
        const len = data.length;
        for (let i = 0; i < len; i++) {
            const v = data[i] ?? null;
            arr.push({
                o: v,
                k: $Call(key_fn, [v, i, data]),
            });
        }
        arr.sort((a, b) => compare(a.k, b.k));
        return arr.map((e) => e.o);
    },
    {
        summary: '根据键函数对数组中的元素进行排序，并返回排序后的结果',
        params: {
            data: '要排序的数组',
            key_fn: '用于提取排序键的函数，接受一个元素并返回其排序键',
            comparator: '用于比较两个排序键的函数，返回一个数字，表示它们的相对顺序，默认按升序排列',
        },
        paramsType: {
            data: 'array',
            key_fn: 'fn(value: any, index: number, arr: type(data)) -> any',
            comparator: 'fn(a: any, b: any) -> number',
        },
        returnsType: 'array',
        examples: ['sort_by([(0, "x"), (2, "y"), (1, "z")], fn (item) { item[0] }) // [(0, "x"), (1, "z"), (2, "y")]'],
    },
);

File: crates/python/mirascript/_vm/lib/vm_global/sequence/sort.py
from __future__ import annotations
from functools import cmp_to_key
import math
from typing_extensions import Callable

from ....._helpers.convert import to_number
from ....._helpers.constants import Uninitialized
from ....operations import Call
from ....types import VmValue, VmAny
from ..._helpers import _expect_array, _expect_callable


def _to_compare(a: int | float, b: int | float) -> int:
    if math.isnan(a):
        a = 0
    if math.isnan(b):
        b = 0
    if a < b:
        return -1
    if a > b:
        return 1
    return 0


def _default_compare(a: VmAny = Uninitialized, b: VmAny = Uninitialized) -> int:
    if a is Uninitialized:
        a = ""
    if b is Uninitialized:
        b = ""
    if isinstance(a, str) and isinstance(b, str):
        if a < b:
            return -1
        elif a > b:
            return 1
        return 0

    if a is b:
        return 0

    return _to_compare(to_number(a, 0), to_number(b, 0))


def _cmp(comparator: VmAny, recovered) -> Callable[[VmValue, VmValue], int]:
    if comparator is None or comparator is Uninitialized:
        return _default_compare
    _expect_callable("comparator", comparator, recovered)

    def compare(a: VmValue, b: VmValue):
        ret = Call(comparator, a, b)
        return _to_compare(to_number(ret), 0)

    return compare


def sort(data=Uninitialized, comparator=Uninitialized):
    data = _expect_array("data", data, None)
    compare = _cmp(comparator, data)
    arr = data.copy()
    arr.sort(key=cmp_to_key(compare))
    return arr


def sort_by(data=Uninitialized, key_fn=Uninitialized, comparator=Uninitialized):
    data = _expect_array("data", data, None)
    key_fn = _expect_callable("key_fn", key_fn, data)
    compare = _cmp(comparator, data)

    arr = []

    l = len(data)

    for i in range(l):
        item = data[i]
        if item is Uninitialized:
            item = None
        key = Call(key_fn, item, i, data)

        arr.append({"o": item, "k": key})

    arr.sort(key=cmp_to_key(lambda a, b: compare(a["k"], b["k"])))
    return [entry["o"] for entry in arr]
